pkg/v13/kubeconfig/kubeconfig.go

```diff
@@ -0,0 +1,129 @@
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/giantswarm/cluster-operator/pkg/v13/key"
+	awskey "github.com/giantswarm/cluster-operator/service/controller/aws/v13/key"
+	azurekey "github.com/giantswarm/cluster-operator/service/controller/azure/v13/key"
+	kvmkey "github.com/giantswarm/cluster-operator/service/controller/kvm/v13/key"
+)
+
+// Config represents the configuration used to create a new index resource.
+type Config struct {
+	// Dependencies.
+	K8sClient kubernetes.Interface
+	Logger    micrologger.Logger
+
+	// Settings.
+	ProjectName       string
+	ResourceName      string
+	ResourceNamespace string
+}
+
+// StateGetter implements the kubeconfig resource.
+type StateGetter struct {
+	// Dependencies.
+	certsSearcher certs.Interface
+	k8sClient     kubernetes.Interface
+	logger        micrologger.Logger
+
+	// Settings.
+	projectName       string
+	resourceName      string
+	resourceNamespace string
+}
+
+// New creates a new configured index resource.
+func New(config Config) (*StateGetter, error) {
+	// Dependencies.
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	// Settings
+	if config.ProjectName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.ProjectName not be empty", config)
+	}
+	if config.ResourceName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.ResourceName must not be empty", config)
+	}
+	if config.ResourceNamespace == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.ResourceNamespace must not be empty", config)
+	}
+
+	var cert certs.Interface
+	{
+		var err error
+		cc := certs.Config{
+			K8sClient: config.K8sClient,
+			Logger:    config.Logger,
+		}
+		cert, err = certs.NewSearcher(cc)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	r := &StateGetter{
+		// Dependencies.
+		certsSearcher: cert,
+		k8sClient:     config.K8sClient,
+		logger:        config.Logger,
+
+		// Settings
+		projectName:       config.ProjectName,
+		resourceName:      config.ResourceName,
+		resourceNamespace: config.ResourceNamespace,
+	}
+
+	return r, nil
+}
+
+func (r *StateGetter) Name() string {
+	return r.resourceName
+}
+
+func ToCustomObject(v interface{}) (*chartconfig.ClusterConfig, error) {
+	var guestConfig v1alpha1.ClusterGuestConfig
+	switch object := v.(type) {
+	case *v1alpha1.AWSClusterConfig:
+		guestConfig = awskey.ClusterGuestConfig(*object)
+	case *v1alpha1.AzureClusterConfig:
+		guestConfig = azurekey.ClusterGuestConfig(*object)
+	case *v1alpha1.KVMClusterConfig:
+		guestConfig = kvmkey.ClusterGuestConfig(*object)
+	default:
+		return nil, microerror.Maskf(invalidConfigError, "cannot identify interface %#v", v)
+	}
```

Comment:
I don't really like this conditional logic. You only need the clusterID and API domain so it could be done like in the `chartconfig` resource.

https://github.com/giantswarm/cluster-operator/blob/0f5052ec451614b689db479a2e859212c54a9350/service/controller/aws/v13/resource/chartconfig/current.go#L37-L40

I'd prefer to have a `kubeconfig` resource for each provider and the shared code in a service in `pkg/v13/kubeconfig`.

For me that structure is easier to understand because the resources are in the usual place.

Reply:
If I follow `chartconfig` resource, then I need to create `ApplyCreateChange`, `ApplyUpdateChange` method and so on again. And that is the opposite direction from what we tried to achieve by generic `secret` resource.

Reply:
To get the Cluster ID and API Domain, anyway we have to transform an instance into cloud cluster config once. doesn't it?

Reply:
No I think the generic secret resource can still be used. Its true you will need `create.go` and `desired.go` in all 3 resources but these can just call the service in `pkg`.

When we flatten cluster-operator to have a single controller we will only need a single resource and the duplication can be removed.

Reply:
Not only `create.go`, `desired.go`, I also need to create `update.go`, `delete.go`, `current.go`. and I don't want to follow that since it's almost we are not using the generic resource.

Reply:
The thing is, if I do wired like this way then each provider controller (aws, azure and kvm) need 6 files (create, update, delete, desired, current) for each. Isn't it correct?

Reply:
I don't see why the create, update etc files are needed. The wiring can be the same as the index resource https://github.com/giantswarm/app-operator/tree/master/service/controller/appcatalog/v1/resource/index.

3 resources are needed which is a consequence of the current cluster-operator design. It will be fixed when we move to a single controller in the Node Pools story.

Reply:
I got what you mean, we can propagate type conversion in actual cloud provider pkg.